mirage/tar_mirage.ml

```diff
@@ -360,33 +397,37 @@ module Make_KV_RW (CLOCK : Mirage_clock.PCLOCK) (BLOCK : Mirage_block.S) = struc
           div data_start_bytes sector_size,
           rem data_start_bytes sector_size
         in
-        let end_sector = div end_bytes sector_size in
-        let end_sector_offset = rem end_bytes sector_size in
+        let end_sector = div (add end_bytes sector_size) sector_size in
+        let last_sector_offset = rem end_bytes sector_size in
         let pad = Tar.Header.compute_zero_padding_length hdr in
 
-        (* Read slack for last sector *)
-        begin
-          if end_sector_offset = 0L then
-            Lwt.return_ok Cstruct.empty
-          else begin
-            let buf = Cstruct.create (to_int sector_size) in
-            BLOCK.read t.b end_sector [buf] >|= function
-            | Error e -> Error (`Block e)
-            | Ok () -> Ok (snd (Cstruct.split buf (to_int end_sector_offset)))
-          end
-        end >>>= fun slack ->
-        let data = Cstruct.concat [ data; Cstruct.create (pad + to_int sentinel); slack ] in
-        let first_sector, remaining_sectors =
-          let s =
-            Stdlib.min
-              (Cstruct.length data)
-              (to_int (sub sector_size data_start_sector_offset))
-          in
-          Cstruct.split data s
+        let data = Cstruct.concat [
+            Cstruct.create (to_int data_start_sector_offset);
+            data;
+            Cstruct.create (pad + to_int sentinel);
+            (if last_sector_offset = 0L then
+               Cstruct.empty
+             else
+               Cstruct.create (to_int (sub sector_size last_sector_offset)));
+          ]
```

Comment:
Here we could as well do a single allocation and blit in the data to reduce the number of allocations.